Add solution for all 4 tasks

Deman4ig opened the pull request:
 Привет!

Вроде сделал.. Слайдер конечно заставил поплакать:)

Commit:
Add solution for all 4 tasks

jsru commented:
Добавляю преподавателя (@SinimaWath) для код-ревью.

SinimaWath commented on 7-module/1-task/index.css:
```diff
@@ -11,7 +11,7 @@
   display: flex;
   flex-direction: row;
-  height: 100%;
+  /*height: 100%;*/
```
Не уверен, это-то что тебе нужно править :)

Deman4ig replied:
На самом деле с этой высотой у меня неправильно лента отображается

<img width="1366" alt="Screen Shot 2021-07-29 at 4 08 44 PM" src="https://user-images.githubusercontent.com/13958183/127559238-48106aec-d99c-4098-9b28-4884ea8d9c07.png">

но это конечно не имеет отношения к коду:)

SinimaWath commented on 7-module/1-task/index.js:
```diff
@@ -1,7 +1,102 @@
+      this._showArrow(this._rightArrow);
+    }
+
+    setTimeout(() => {
```
немного не понимаю зачем тут setTimeout - просто попробовать?

Deman4ig replied:
в задании говорилось что надо скрывать кнопки когда сам скролл закончился.. тк скролл асинхронный и поэтому добавил таймаут.. лучше ничего в голову не пришло:)

SinimaWath commented on 7-module/2-task/index.js:
```diff
@@ -1,6 +1,71 @@
+    document.body.append(this._modal);
+
+    this._closeBtn.addEventListener('click', this.close, { once: true });
+    document.body.addEventListener('keydown', this._closeOnEsc);
```
Тут тоже надо очищать keydown, иначе на каждый open, создается подписка, но не удаляется

Deman4ig replied:
согласен! в слайдере уже понял что надо удалять:)

Commit:
Refactor task-1 and task-2

task-1: css
task-2: remove event from body on modal close

jsru commented:
Решение было обновлено, посмотрим что скажет @SinimaWath

SinimaWath commented on 7-module/1-task/index.js:
```diff
@@ -1,7 +1,102 @@
+}
+
 export default class RibbonMenu {
+  _ribbon;
+  _leftArrow;
+  _rightArrow;
+  _ribbonInner;
```
Надо задать значение по-умолчанию свойствам

Deman4ig replied:
Пересмотрел видео, где ты рассказывал про классы и поставил `prop = null`

SinimaWath commented on 7-module/3-task/index.js:
```diff
@@ -1,4 +1,100 @@
+import createElement from '../../assets/lib/create-element.js';
+
+function generateSliderHTML({ steps, value }) {
+  const div = document.createElement('div');
+
+  for (let i = 0; i < steps; i += 1) {
+    const span = document.createElement('span');
+
+    if (i === value) {
+      span.classList.add('slider__step-active');
+    }
+
+    div.append(span);
+  }
```
Тебе тут не нужен временный div.
Сейчас ты делаешь так: html -> div -> html.

Можно просто сразу строку формировать без временного div
 Попробуй через map + join

Deman4ig replied:
Согласен с первой частью коммента, но не уверен что моя реализация через map + join это то что ты имел ввиду

SinimaWath commented on 7-module/3-task/index.js:
```diff
@@ -1,4 +1,100 @@
+}
+
 export default class StepSlider {
+  _slider;
+  _steps;
+  _value;
+  _segments;
```
все таки лучше явно задавать значения по-умолчанию

Deman4ig replied:
Пересмотрел видео, где ты рассказывал про классы и поставил `prop = null`

SinimaWath commented on 7-module/3-task/index.js:
```diff
@@ -1,4 +1,100 @@
+    let highlightedElement = this._slider.querySelector('.slider__step-active');
+    highlightedElement.classList.remove('slider__step-active');
+    this._sliderValue.innerText = value;
+    this._slider.querySelector(`.slider__steps > span:nth-of-type(${value + 1})`).classList.add('slider__step-active');
```
Тут понятней будет querySelectorAll и брать по индексу

Deman4ig replied:
была такая мысль изначально, но подумал что может с точки зрения использования памяти найти конкретный див оптимальнее чем сложить массив в переменную?

Попробовал сделать через массив.

Commit:
Refactor

1. Add default values to class properties
2. Task 3 - update slider value using elem.querySelectorAll
3. Task 3 - update function to build slider template

jsru commented:
Добавляю преподавателя (@SinimaWath) для код-ревью.

SinimaWath reviewed (approved):
Все круто :)
